### PR TITLE
Migrate Geti object detection models

### PR DESCRIPTION
**Pull request by jaegukhyun**
### Summary
Migrate Geti object detection models.
Learning hyperparameters could be changed after this PR is merged.
Tests is needed but it will be added in next PR.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/__init__.py [hunk 1]
```diff
@@ -2,3 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 """Module for OTX custom algorithms, e.g., model, losses, hook, etc..."""
+
+from . import classification, detection
+
+__all__ = ["classification", "detection"]
```
Comment (vinnamkim) [inline src/otx/algo/__init__.py:8]: I don't like this since `import otx.algo` can be slow to import everything. This might impact our CLI entrypoint performance if not managed well. If you need registration for `mmx.Registry`, it seems better to make a some helper function for each task.

File: src/otx/core/model/entity/detection.py [hunk 2]
```diff
@@ -105,7 +104,13 @@ def _customize_outputs(
 
             losses = OTXBatchLossEntity()
             for k, v in outputs.items():
-                losses[k] = sum(v)
+                if isinstance(v, list):
+                    losses[k] = sum(v)
+                elif isinstance(v, torch.Tensor):
+                    losses[k] = v
+                else:
+                    message = "Loss output should be list or torch.tensor but got {type(v)}"
```
Comment (vinnamkim) [inline src/otx/core/model/entity/detection.py]: I think it will be good to follow the previous convention `msg="..."` rather than making a new one for the future maintenance.

File: src/otx/core/model/module/detection.py [hunk 3]
```diff
@@ -59,6 +59,9 @@ def _log_metrics(self, meter: MeanAveragePrecision, key: str) -> None:
                 log.debug("Cannot log Tensor which is not scalar")
                 continue
 
+            if k != "map_50":
```
Comment (vinnamkim) [inline src/otx/core/model/module/detection.py]: What's the reason this metric should be dropped?

File: src/otx/core/utils/build.py [hunk 4]
```diff
@@ -18,6 +16,11 @@
 
 def build_mm_model(config: DictConfig, model_registry: Registry, load_from: str) -> nn.Module:
     """Build a model by using the registry."""
+    from mmengine.runner import load_checkpoint
+
+    from otx import algo  # noqa: F401
```
Comment (vinnamkim) [inline src/otx/core/utils/build.py:21]: Please see my above comment

File: src/otx/recipe/detection/atss_mobilenetv2.yaml [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    val:
+      batch_size: 1
+      transforms:
+        - type: LoadImageFromFile
+        - type: Resize
+          scale:
+            - 992
+            - 736
+          keep_ratio: false
+        - type: LoadAnnotations
```
Comment (vinnamkim) [inline src/otx/recipe/detection/atss_mobilenetv2.yaml:50]: How did you solved the problem you wrote in the OneNote page. I saw that there exists a resizing problem for the ground truth annotations.

File: src/otx/core/data/module.py [hunk 6]
```diff
@@ -62,10 +62,8 @@ def _get_config(self, subset: str) -> SubsetConfig:
 
     def _get_dataset(self, subset: str) -> OTXDataset:
         if (dataset := self.subsets.get(subset)) is None:
-            msg = (
-                f"Dataset has no '{subset}'. Available subsets = {self.subsets.keys()}"
-            )
-            raise KeyError(msg)
+            message = f"Dataset has no '{subset}'. Available subsets = {list(self.subsets.keys())}"
```
Comment (vinnamkim) [inline src/otx/core/data/module.py]: Same here

File: src/otx/core/model/module/detection.py (hunk 3 above)
Comment (jaegukhyun) [reply src/otx/core/model/module/detection.py]: Other metrics, such as mAP_75, mAP_large, mAP_small, etc, make log messy. I think reporting just mAP50 is fine.

File: src/otx/recipe/detection/atss_mobilenetv2.yaml (hunk 5 above)
Comment (jaegukhyun) [reply src/otx/recipe/detection/atss_mobilenetv2.yaml:50]: The problem happens when Resize after LoadAnnotation. Therefore I changed all detection configs to Resize before LoadAnnotation.

File: src/otx/algo/__init__.py (hunk 1 above)
Comment (jaegukhyun) [reply src/otx/algo/__init__.py:8]: Making helper function looks good to me. I'll make a separate PR for that.

File: src/otx/core/model/entity/detection.py (hunk 2 above)
Comment (jaegukhyun) [reply src/otx/core/model/entity/detection.py]: Done by ed75b5761d6ca2bcf69c8bcf14f01e69c37d1eb3

File: src/otx/core/data/module.py (hunk 6 above)
Comment (jaegukhyun) [reply src/otx/core/data/module.py]: Done by ed75b5761d6ca2bcf69c8bcf14f01e69c37d1eb3

File: src/otx/core/model/module/detection.py (hunk 3 above)
Comment (vinnamkim) [reply src/otx/core/model/module/detection.py]: That will affect to the logger behavior not just CLI stdout (progress bar). If you don't want too verbose, I recommend
```python
            self.log(
                f"{key}/{k}",
                v,
                sync_dist=True,
                prog_bar=True if k == "map_50" else False,
            )
```

File: src/otx/recipe/detection/atss_mobilenetv2.yaml (hunk 5 above)
Comment (vinnamkim) [reply src/otx/recipe/detection/atss_mobilenetv2.yaml:50]: You mean that your model can resize it's bbox output to the original image scale, right?

Comment (jaegukhyun) [reply src/otx/recipe/detection/atss_mobilenetv2.yaml:50]: Yes, mmdet's predict mode returns prediction of original image size

File: src/otx/core/model/module/detection.py (hunk 3 above)
Comment (jaegukhyun) [reply src/otx/core/model/module/detection.py]: Reflected by [eb22d3e](https://github.com/openvinotoolkit/training_extensions/pull/2690/commits/eb22d3e4be286319f8433fc44c02ada8d8d49f53)

**Review by vinnamkim — APPROVED** [review]
LGTM.

File: src/otx/core/model/module/detection.py (hunk 3 above)
Comment (vinnamkim) [reply src/otx/core/model/module/detection.py]: In the future PR, I think we can make it configurable such as `k in self.prog_bar_metrics` and `self.prog_bar_metrics = {"map_ap"}`.

File: src/otx/recipe/detection/atss_mobilenetv2.yaml [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            - scale_factor
+            - gt_ann_ids
+            - flip
+            - ignored_labels
+            - ori_shape
+            - filename
+            - img_shape
+            - pad_shape
+    val:
+      batch_size: 1
```
Comment (kprokofi) [inline src/otx/recipe/detection/atss_mobilenetv2.yaml:42]: Why do we use batch size 1 for validation?

**Review by kprokofi — APPROVED** [review]
Overall looks good to me. Creating function for the model registry is a good option.

**Review by sungmanc — APPROVED** [review]
LGTM, but we need to unify the docstring format. docstring --> license ? or lincense --> docstring?

File: src/otx/recipe/detection/atss_mobilenetv2.yaml (hunk 7 above)
Comment (jaegukhyun) [reply src/otx/recipe/detection/atss_mobilenetv2.yaml:42]: In OTX1.5 Some issues happen occasionally when validation batch size is over 1. I'll check it later and update this
Thanks for the comment
